images/metering-ansible-operator/roles/meteringconfig/library/k8s_event.py

```diff
@@ -0,0 +1,352 @@
+
+        involved_obj = self.params.get("involvedObject")
+        print(involved_obj["kind"],involved_obj["namespace"],involved_obj["name"],involved_obj["apiVersion"])
+
+
+        resource = self.find_resource(
+            "Event", "v1", fail=True)
+
+        priorEvent = None
+        try:
+            priorEvent = resource.get(
+                name=meta["name"],
+                namespace=meta["namespace"])
+
+        except openshift.dynamic.exceptions.NotFoundError:
+            pass
+
+        # event counter
+        priorCount = 1
+        print("the count is no event exists %i" % priorCount)
+        now = datetime.datetime.now()
+        rfc = kubernetes.config.dateutil.format_rfc3339(now)
+        firstTimestamp = rfc
+        lastTimestamp = rfc
+
+        if priorEvent:
+            if priorEvent["reason"] != reason:
+                print("If event exists and reason is changed %i" % priorCount)
+                now = datetime.datetime.now()
+                rfc = kubernetes.config.dateutil.format_rfc3339(now)
+                priorCount = 1
+                firstTimestamp = rfc
+                print("the value of the firstTimestamp new event is", firstTimestamp)
+                lastTimestamp = rfc
+                print("the value of the lastTimestamp new event is", lastTimestamp)
+            else:
+                priorCount = priorEvent["count"] + 1
+                print("If event exists and reason is the same %i" % priorCount)
+                firstTimestamp = priorEvent["firstTimestamp"]
+                print("the value of the firstTimestamp event exists same reason is", firstTimestamp)
+                lastTimestamp = firstTimestamp
+                print("the value of the lastTimestamp event exists same reason is", lastTimestamp)
+
+        # handle involvedObject
+        involvedObject_resourceVersion = None
+        involvedObject_uid = None
+        involvedObjectResource = self.find_resource(
+            involved_obj["kind"], "v1", fail=True)
+
+        if involvedObjectResource:
+            try:
+                apiInvolvedObject = resource.get(
+                    name=involved_obj["name"], namespace=involved_obj["namespace"])
+
+                involvedObject_uid = apiInvolvedObject["metadata"]["uid"]
+                involvedObject_resourceVersion = apiInvolvedObject["metadata"]["resourceVersion"]
+
+            except openshift.dynamic.exceptions.NotFoundError:
+                pass
+
+        event = {
+            "apiVersion": "v1",
+            "count": priorCount,
+            "eventTime": None,
+            "firstTimestamp": firstTimestamp,
+            "involvedObject": {
+                "apiVersion": involved_obj["apiVersion"],
+                "kind": involved_obj["kind"],
+                "namespace": involved_obj["namespace"],
+                "name": involved_obj["name"],
+                "resourceVersion": involvedObject_resourceVersion,
```

Comment:
It looks like you would need to validate that this/these value(s) aren't `None` as you initialize it in a try/except block, and there could be an error attempting to get the involved_object, which means that these values contain their default, assigned values (`None`).

Reply:
It's there. See line 298

Reply:
On line 292, you initializing the variable to `None`:
```python
involvedObject_resourceVersion = None
```
If `involvedObjectResource` evaluates to false, or there's an error while running:
```python
try:
                apiInvolvedObject = resource.get(
                    name=involved_obj["name"], namespace=involved_obj["namespace"])
```
then `involvedObject_resourceVersion` won't be re-initialized, and you would referencing the original value.

I'm not sure what happens if you pass `"resourceVersion": involvedObject_resourceVersion,`
to the event dictionary when `involvedObject_resourceVersion == None` and whether or not that would produce an error.

Reply:
At every major code change I test:
New Event,
New Event /same reason
New Event /different reason
Deletion and re-creation of event
I have not hit an error in any of these.

Reply:
@timflannagan1 that's expected. These values will be unset if the involvedObject doesn't exist since we can't get them, but we still want to create the event, even if we can't get all the details of the involvedObject.